x/auth/middleware/middleware.go

```diff
@@ -102,11 +116,14 @@ func NewDefaultTxHandler(options TxHandlerOptions) (tx.Handler, error) {
 		SigGasConsumeMiddleware(options.AccountKeeper, options.SigGasConsumer),
 		SigVerificationMiddleware(options.AccountKeeper, options.SignModeHandler),
 		IncrementSequenceMiddleware(options.AccountKeeper),
+		endAnteBranch,
+
 		// Creates a new MultiStore branch, discards downstream writes if the downstream returns error.
 		// These kinds of middlewares should be put under this:
 		// - Could return error after messages executed succesfully.
-		// - Storage writes should be discarded together when tx failed.
-		WithBranchedStore,
+		// - Storage writes below this middleware should ALWAYS be discarded
+		//   together when middleware fails, but those above can have state writes.
+		WithRunMsgsBranch,
 		// Consume block gas. All middlewares whose gas consumption after their `next` handler
 		// should be accounted for, should go below this middleware.
 		ConsumeBlockGasMiddleware,
```

Comment:
This `ConsumeBlockGasMiddleware` is a new middleware added in v0.46, which afair differs from v0.45 logic. https://github.com/cosmos/cosmos-sdk/pull/10770.

I would love @yihuang (or someone else) to double-check the current PR still does what #10770 intended.

Reply:
I thought the middleware was just a refactored version of the ante handler, with particular semantics related to checkTx/deliverTx.
I had some understanding of how ante handlers work, but I'm not sure how the middleware are supposed to work right now.

Reply:
looks alright, the `WithRunMsgsBranch` is just renamed `WithBranchedStore` right?

Reply:
would it be clearer if we keep two separate middleware stacks for ante handlers and runMsgs, and they are run under different branched stores (db transaction).

Reply:
@yihuang I think there are multiple people saying that middlewares are confusing. I created https://github.com/cosmos/cosmos-sdk/issues/11955